### PR TITLE
[cli] Toggle Primary Instance (petenv)

### PR DESCRIPTION
**Pull request by surahman**
**_Issue #1138:_**
I have added support to handle the `client.primary-name=` command. I am not sure about the mock tests I have set up and they would require some review. When stepping through the tests with a debugger I was unable to hit breakpoints on `set` and `set_aux` within the `utils/settings.cpp` file - which is typically an indicator of an issue. Manual argument passing tests work on the `cmdl`.

### CODE REVIEW

File: tests/test_mock_settings.cpp
Comment (ricab) [inline tests/test_mock_settings.cpp:65]: The situation with `Settings` and tests is particular... We don't really have tests for the real `Settings` singleton, because it relies on `QSettings` which is not mockable. We'd have to provide another wrapper for `QSettings` which we could extend and mock and we never got around to that. 

OTH, we don't want tests to affect or be affected by real settings on the disk, so we make sure [all the tests run with a `MockSettings`](https://github.com/canonical/multipass/blob/c3d3afa0d62b9e3a766a769bd72d3d5f4ea84ac2/tests/main.cpp#L36) instead. This file has only a couple of very shallow tests for the mock itself, to show that it works when accessing the `Settings` singleton.

In short, what you wrote here only calls a mock and we don't have a way to actually test this ATM, so you can remove these entirely.

File: src/utils/settings.cpp
Comment (ricab) [inline src/utils/settings.cpp:211]: Hmm, I don't think this is right. You want to accept empty values, so you should just add that condition to the first `if`:

```c++
if (key == petenv_key && !val.isEmpty() && !mp::utils::valid_hostname(val.toStdString()))
    throw ...
```

Note that `petenv_key` is the constant defining `"client.primary-name"` as the key we use for that.

File: src/client/cli/cmd/set.cpp
Comment (ricab) [inline src/client/cli/cmd/set.cpp:128]: It appears you added some stray newlines (in an otherwise empty commit).

**Review by ricab — CHANGES_REQUESTED** [review]
Hi @surahman, thanks for the PR. I am afraid it doesn't cut it yet though. 

The logic for disabling primary would need to go in the commands that deal with it (along with the "if" condition below).

File: tests/test_mock_settings.cpp
Comment (surahman) [reply tests/test_mock_settings.cpp:65]: Got it, thanks for the explanation!

File: src/utils/settings.cpp
Comment (surahman) [reply src/utils/settings.cpp:211]: Corrected, I had not looked up the `constants` prior.

File: src/client/cli/cmd/set.cpp
Comment (surahman) [reply src/client/cli/cmd/set.cpp:128]: Cleaned up. The only intended change here was to silence a deprecated namespace warning.

**Comment by surahman** [comment]
> The logic for disabling primary would need to go in the commands that deal with it (along with the "if" condition below).
I will correct the issue with the error being thrown about the missing settings key parameter.

**Comment by ricab** [comment]
Hey @surahman, yeah I mean the places where the primary instance gets special treatment:

```
$ grep -r petenv_key src/client/ include/
src/client/gui/gui_cmd.cpp:    auto petenv_name = MP_SETTINGS.get(petenv_key).toStdString();
src/client/cli/cmd/restart.cpp:    const auto petenv_name = MP_SETTINGS.get(petenv_key);
src/client/cli/cmd/stop.cpp:    const auto petenv_name = MP_SETTINGS.get(petenv_key);
src/client/cli/cmd/suspend.cpp:    const auto petenv_name = MP_SETTINGS.get(petenv_key);
src/client/cli/cmd/launch.cpp:    petenv_name = MP_SETTINGS.get(petenv_key);
src/client/cli/cmd/shell.cpp:    petenv_name = MP_SETTINGS.get(petenv_key);
src/client/cli/cmd/start.cpp:    petenv_name = MP_SETTINGS.get(petenv_key);
include/multipass/constants.h:constexpr auto petenv_key = "client.primary-name";     // This will eventually be moved to some dynamic settings schema
include/multipass/cli/format_utils.h:    const auto petenv_name = MP_SETTINGS.get(petenv_key).toStdString();
```

They would have to be modified to have the primary instance disabled. Otherwise this just sets the name to the empty string. 

(Note that, for historic reasons, the code often uses the term `petenv`, short for "pet environment", to designate the primary instance.)

File: src/client/cli/cmd/set.cpp
Comment (ricab) [reply src/client/cli/cmd/set.cpp]: The problem is that doesn't work with earlier Qt versions:

```
./src/client/cli/cmd/set.cpp:126:59: error: ‘KeepEmptyParts’ is not a member of ‘Qt’
  126 |             const auto keyval = args.at(0).split('=', Qt::KeepEmptyParts);
      |                                                           ^~~~~~~~~~~~~~
```

So we decided to live with these warnings for now.

File: src/utils/settings.cpp
Comment (ricab) [inline src/utils/settings.cpp:205]: OK, so only this change for now. Would you mind squashing all the commits so far accordingly? Thanks.

**Comment by surahman** [comment]
> ```
> $ grep -r petenv_key src/client/ include/
> src/client/gui/gui_cmd.cpp:    auto petenv_name = MP_SETTINGS.get(petenv_key).toStdString();
> src/client/cli/cmd/restart.cpp:    const auto petenv_name = MP_SETTINGS.get(petenv_key);
> src/client/cli/cmd/stop.cpp:    const auto petenv_name = MP_SETTINGS.get(petenv_key);
> src/client/cli/cmd/suspend.cpp:    const auto petenv_name = MP_SETTINGS.get(petenv_key);
> src/client/cli/cmd/launch.cpp:    petenv_name = MP_SETTINGS.get(petenv_key);
> src/client/cli/cmd/shell.cpp:    petenv_name = MP_SETTINGS.get(petenv_key);
> src/client/cli/cmd/start.cpp:    petenv_name = MP_SETTINGS.get(petenv_key);
> include/multipass/constants.h:constexpr auto petenv_key = "client.primary-name";     // This will eventually be moved to some dynamic settings schema
> include/multipass/cli/format_utils.h:    const auto petenv_name = MP_SETTINGS.get(petenv_key).toStdString();
> ```

@ricab thanks for the information. How would you like to have the `primary` is disabled situation dealt with? Is the correct of dealing with this to just return from the function with a `success` (potentially a silent error) or a `failure` (input command/request invalid) code?

**Comment by ricab** [comment]
Hi @surahman, this is quite tricky. if the primary instance is disabled, multipass needs to disable any special treatment for it and error out in commands that need an instance but get none. So, if the user said for instance `multipass start`, without an instance name, multipass would indeed need to error out. Commands like start, stop, shell... would require at least one instance argument. Pretty much like a `multipass delete` would, but perhaps mentioning that the primary instance is disabled in the error message. The `help` messages would also need to be updated, where they mention primary today.

**Comment by ricab** [comment]
BTW, I haven't looked at your new code yet. Please let me know when you feel it's ready for another review pass.

**Comment by surahman** [comment]
@ricab I have gone ahead and updated the tests to include parameters. ATM `start` errors out if a command is passed without a parameter.

The other commands are also set to error out if the `petenv` is disabled. I gather that you would like a message to be sent to `cerr` and the command to return `ok` instead? Are there any example tests on how to capture the messages sent to the streams?

I will look through the `help` messages to locate instances of `primary`.

**Comment by ricab** [comment]
Hmm, I don't quite follow, but primary is an instance that is automatically launched, started, stoped, etc. when you use the corresponding commands without arguments. And it's name can change, so that after `multipass set client.primary_name=heyo`, `multipass start # no args` launches an instance named `heyo`, `multipass stop` would stop it, etc.  BTW, have you seen https://multipass.run/docs/primary-instance ?

So, to disable it, `stop`, `start`, etc. should now error out, since they don't know what instance to work with. So that's what the tests should check, for instance:

```c++
    EXPECT_CALL(mock_settings, get(Eq(mp::petenv_key))).WillRepeatedly(Return(""));
    EXPECT_THAT(send_command({"start" /* no further args */ }), Eq(mp::ReturnCode::CommandFail));
```

while
```c++
    EXPECT_CALL(mock_settings, get(Eq(mp::petenv_key))).WillRepeatedly(Return("something"));
    EXPECT_THAT(send_command({"start" /* no further args */ }), Eq(mp::ReturnCode::CommandSuccess));
```

The `launch` command is different, because it expects an image name rather than an instance name. So `multipass launch # no args` is unrelated to primary.

BTW, the help messages are dynamically generated, so you should look for `petenv_key` rather than `primary`.

**Comment by surahman** [comment]
@ricab I think it is ready for a review. The handling is mostly complete except for the cases where the `primary` is disabled and we have flags such as `--all`. As it stands right now, this will cause an error, even if there are other active instances with the disabled `primary`.

**Comment by ricab** [comment]
> (Jarjar Binks is happy)

Great to hear that @surahman :wink: 

Thank you again for this! I will have to queue reviews for some days to advance on other fronts, but I want to keep answering your questions if you have any in the meantime. So keep asking away and I'll reply as I can.

> For the others, I will need to handle the disabled primary in the parse_args to check for additional passed parameter counts when necessary. Still drop an error message on cerr?

> [...] except for the cases where the primary is disabled and we have flags such as --all

I am afraid I still don't entirely understand the question, so I'll give a broader reply in the hope that it covers it. 

From the top of my head, the only particularity of `start` (along with `shell`), with respect to the primary instance, is that it launches it if it does not exist. This is not applicable to `restart` as things stand today because only running instances are restartable. IIRC the only way `launch` treats the primary instance differently is by mounting the home folder. 

Other than that, the primary instance is just a default argument when no instance and no `--all` is passed. To disable primary, we need to refuse `start`, `stop`, `suspend`, etc. with no arguments and no `--all`. 

Note that an instance with the name "primary" (and whatever `client.primary_name` was before disabling) can still exist, be launched, stopped, and otherwise function normally, like any other instance. That's no different from what happens today when you have an instance named "primary" and change the primary name to, say, "foo". If you say `multipass stop`, "foo" will be stopped, "primary" will be untouched. And `--all` just keeps targeting all instances. 

Here is an expected interaction (messages not accurate):

```
$ multipass list
No instances
$ multipass stop --all
[NOOP]
$ multipass start --all
[NOOP]
$ multipass start
[launches primary, auto-mounts]
$ multipass set client.primary-name=foo
$ multipass stop
error - foo does not exist
$ multipass stop --all
[stops primary]
$ multipass start
[launches foo]
$ multipass start --all
[starts primary]
$ multipass delete -p primary
[deletes primary]
$ multipass launch -n primary
[launches primary, no auto-mount]
$ multipass set client.primary-name=
$ multipass stop
error: primary instance disabled, need an instance name or --all
$ multipass start
error: primary instance disabled, need an instance name or --all
$ multipass stop primary
[stops primary]
$ multipass stop foo
[stops foo]
$ multipass start --all
[starts foo and primary]
$ multipass delete --all --purge
[deletes foo and primary]
$ multipass launch -n primary
[launches an instance named primary, no auto-mount]
$ multipass launch -n foo
[launches an instance named foo, no auto-mount]
$ multipass set client.primary-name=primary
$ multipass stop
[stops primary]
```

**Comment by surahman** [comment]
Thanks, I am trying to understand the valid `CLI` parameters without any experience with the `CLI`, so I am relying on documentation and tests like these:
```c++
TEST_F(Client, start_cmd_considers_configured_petenv)
{
    const auto custom_petenv = "jarjar binks";
    EXPECT_CALL(mock_settings, get(Eq(mp::petenv_key))).WillRepeatedly(Return(custom_petenv));

    const auto petenv_matcher = make_instance_in_repeated_field_matcher<mp::StartRequest, 1>(custom_petenv);
    EXPECT_CALL(mock_daemon, start(_, petenv_matcher, _));
    EXPECT_THAT(send_command({"start"}), Eq(mp::ReturnCode::Ok));
}
```

**Comment by surahman** [comment]
> $ multipass set client.primary-name=
> $ multipass start
> error: primary instance disabled, need an instance name or --all

@ricab focusing on `start` here, I have the checks for passing the `--all` flag working but I now need to employ a check to see if other instances are available. Do I need to send an IPC `dispatch` request for a `list` to obtain a list of available instances, or is there some pre-built routine available to that end?

**Comment by surahman** [comment]
Sorry, just renamed the branch it closed the link to this PR.

**Comment by ricab** [comment]
Hi @surahman, why would you need that "check to see if other instances are available"? If you did, then having a common function would make sense, but I don't see why you would. A start command with no arguments will be wrong independently of what instances are available at the moment.

There is actually a common function to check the validity of the command in this respect, so you could drop [this if](https://github.com/canonical/multipass/pull/2102/files#diff-2868af61f453cf60edb3c228b594b5121226ece5e07637e0d4a3b9abc804e6faR179) and instead edit the call just below:

```diff
diff --git a/src/client/cli/cmd/start.cpp b/src/client/cli/cmd/start.cpp
index 1c94ecb..7609f09 100644
--- a/src/client/cli/cmd/start.cpp
+++ b/src/client/cli/cmd/start.cpp
@@ -173,7 +173,7 @@ mp::ParseCode cmd::Start::parse_args(mp::ArgParser* parser)
     if (status != ParseCode::Ok)
         return status;
 
-    auto parse_code = check_for_name_and_all_option_conflict(parser, cerr, /*allow_empty=*/true);
+    auto parse_code = check_for_name_and_all_option_conflict(parser, cerr, /*allow_empty=*/ !petenv_name.isEmpty());
     if (parse_code != ParseCode::Ok)
         return parse_code;
```

**Comment by surahman** [comment]
> Hi @surahman, why would you need that "check to see if other instances are available"?

Sorry, my modifications were erroring out `start -all` and I thought there needed to be a check to see if there are any available instances when the `--all` flag is passed which can actually be launched. Thanks for the clarification!

> ```diff
> diff --git a/src/client/cli/cmd/start.cpp b/src/client/cli/cmd/start.cpp
> index 1c94ecb..7609f09 100644
> --- a/src/client/cli/cmd/start.cpp
> +++ b/src/client/cli/cmd/start.cpp
> @@ -173,7 +173,7 @@ mp::ParseCode cmd::Start::parse_args(mp::ArgParser* parser)
>      if (status != ParseCode::Ok)
>          return status;
>  
> -    auto parse_code = check_for_name_and_all_option_conflict(parser, cerr, /*allow_empty=*/true);
> +    auto parse_code = check_for_name_and_all_option_conflict(parser, cerr, /*allow_empty=*/ !petenv_name.isEmpty());
>      if (parse_code != ParseCode::Ok)
>          return parse_code;
> ```

Thanks, this tightens up the code.

**Comment by surahman** [comment]
I have made adjustments, all tests are passing locally. I am not sure if the failed build pipeline is due to the branch rename and subsequent closing.

**Comment by surahman** [comment]
@townsend2010 thanks for the info :)

**Review by ricab — CHANGES_REQUESTED** [review]
Alright, I finally managed to circle back to this, sorry for the delay here.

This is getting closer, but a few aspects still need improving and one important piece is still missing: the GUI. I had to skip the tests, leaving that for a later pass.

Most of my requests are repeated in the multiple affected commands. Note that, when I mentioned updating the help messages, I meant "where they mention primary today", so mainly `addPositionalArgument` AFAICS, not the general description.

File: include/multipass/cli/command.h
Comment (surahman) [reply include/multipass/cli/command.h:138]: Thanks for the feedback, I will look into these issues as time permits. I have not spent all that much time reviewing this code, just scattered work on it when I had time.

**Comment by surahman** [comment]
I hope I got everything from this review cycle 🤞🏽.

**Comment by surahman** [comment]
Found some time and reviewed your comments again, please let me know if anything got overlooked.

**Comment by surahman** [comment]
If the CLI updates are complete we can look at the GUI. What is supposed to be happening there?

I would like to complete the two PR I have open before moving on to others.

**Comment by surahman** [comment]
A quick attempt at wiring in the GUI. This is the first time I am running the build on any of my machines and I am running into issues where the GUI does not seem to be able to open the shell, installing from snap does not have this issue. Not sure if I have broken something in the build or if I am missing instructions (followed readme.md).

**Comment by surahman** [comment]
A quick attempt at wiring in the GUI in some spare time. This is the first time I am running the GUI of a build on any of my machines and I am running into issues where the GUI does not seem to be able to open the shell. Installing from snap does not have this issue. Not sure if I have broken something in the build or if I am missing instructions (following `README.md`).

When disabling the primary instance we may need to stop it, which I believe would need to be wired into the `create_actions`
here:
https://github.com/surahman/multipass/blob/fb4e4559c7a14c0ba354451cabb769ac3d8b289c/src/client/gui/gui_cmd.cpp#L179

**Review by ricab — COMMENTED** [review]
Nice nice, this is getting close :slightly_smiling_face:

The CLI part is practically done. The GUI that I am less sure of. I think for the time being it would be better to just disable (grey out) the 3 buttons concerning primary. But I will discuss this with others and give you some more feedback. BTW, the start button is missing when launching the GUI with a disabled primary.

The [help for the name option in launch](https://github.com/canonical/multipass/blob/bd340e8d9703fafcf3349bdf05fe1804b7ec7e88/src/client/cli/cmd/launch.cpp#L199) also needs updating (it shouldn't mention primary or auto-mounting when disabled).

File: src/client/gui/gui_cmd.cpp
Comment (ricab) [reply src/client/gui/gui_cmd.cpp:64]: Or maybe just an enable button (but no disable). We don't need to worry about maintaining the previous name if it was reset elsewhere.

Comment (ricab) [reply src/client/gui/gui_cmd.cpp:64]: Hey @surahman, we have a release to ship and folks are on holidays, so we'll need to defer discussion until later and we'd better put the GUI part on hold.

Comment (surahman) [reply src/client/gui/gui_cmd.cpp:64]: > we have a release to ship and folks are on holidays

@ricab all work and no play make life dull 😉.

I have begun resolving the `CLI` issues and started refactoring and updating the tests.

File: tests/test_cli_client.cpp
Comment (surahman) [reply tests/test_cli_client.cpp]: The purpose of this test is to ensure the changes I have made do not break the `--all` option. I have removed it.

**Review by ricab — COMMENTED** [review]
Looking good, thanks @surahman! 

We'll get back to you with a path for the GUI soon.

**Comment by surahman** [comment]
@ricab no worries.

I would ask that you folks consider the following scenario:
Primary Instance is running.
User disables Primary Instance.
No checks or shutdown signal is sent to the Primary Instance, it continues to run.
User re-enables Primary Instance.

In the best case scenario, the Primary Instance name is stored and the user can issue commands to it again. Worst case scenario, if you assign a new random name using the functions in `src/petname/`, you have an orphaned Primary Instance. This is why I wanted to check for an active Primary Instance on the daemon via an RPC or some other method.

**Comment by ricab** [comment]
Hey @surahman, I think that's fine. It's similar to changing the name to something else. Suppose: 

1. a primary instance exists, with name "primary"
2. the user does `multipass set client.primary-name=another`

The instance named "primary" remains, but it becomes a regular instance that happens to be named "primary". A `multipass start` at that point would launch "another". Setting to "" just makes it so that `multipass start` would instead fail.

So, in both cases:

3. `multipass set client.primary-name=primary`

Just makes the instance named "primary" get the special treatment again.

Notice we use the term "primary" as both instance property and instance name. Perhaps a little confusing, but just realize an instance named "primary" doesn't need to have the primary property, and vice-versa.

**Comment by ricab** [comment]
@surahman for the GUI, we discussed a bit and we think the best for now is to just remove the entire primary section: buttons and separator. They would need to be added back when re-enabled (reflecting the current state).

Additionally, the [hotkey](https://github.com/canonical/multipass/blob/main/src/client/gui/gui_cmd.cpp#L107) should not open a shell anymore. That may be easier to achieve by just checking if primary is disabled inside the action callback.

**Comment by surahman** [comment]
@ricab thanks for all the info - I will get started on this, just waiting on the formatter pull to be complete. The delta between the code means I have to discard all the build artifacts and do a clean build, and that takes a while...

**Comment by Saviq** [comment]
> @ricab thanks for all the info - I will get started on this, just waiting on the formatter pull to be complete. The delta between the code means I have to discard all the build artifacts and do a clean build, and that takes a while...

That shouldn't be necessary? Just building again should rebuild just the needed bits. I can also recommend `ccache`, and well - you could always have multiple build dirs :)

**Comment by surahman** [comment]
@Saviq thank you, I will change my workflows. I currently have a single `build` directory and I `stash` my `gitignore`, and then `stash` my build artefacts, then switch branches, then pop my `stash` and go on my merry way.

This time I jumped from the `cloud-init-iso` branch to this old branch and got a whole load of undefined reference errors for `adapter-something-something` during linking. My guess is `make` compared the timestamps of last modified on the compilation units against their artefacts and decided they did not need updating.

**Comment by ricab** [comment]
Hey @surahman, you should be able to add your custom ignore stuff to `.git/info/exclude` in your local repo, instead of stashing `.gitignore`. You can also have your own global `.gitignore`, if that makes sense for your other projects.

I also just leave the build tree around, only clearing it if there are changes to CMake that require it (my IDE manages that sort of thing actually). And yeah, `ccache` is a must. Anyway, take your time of course :slightly_smiling_face:

**Comment by surahman** [comment]
> Hey @surahman, you should be able to add your custom ignore stuff to `.git/info/exclude` in your local repo, instead of stashing `.gitignore`. You can also have your own global `.gitignore`, if that makes sense for your other projects.

@ricab thank you! This makes life a lot easier 😄, and I shall get `ccache` installed as well.

> just remove the entire primary section: buttons and separator. They would need to be added back when re-enabled (reflecting the current state).

I wanted to clarify that you want me to remove the complete `primary` section, except a button to re-enable it, when the `disable primary` is toggled on, and returned when it is toggled off?

**Comment by ricab** [comment]
> @ricab thank you! This makes life a lot easier smile, and I shall get ccache installed as well.

Sure thing! Glad to help :)

> I wanted to clarify that you want me to remove the complete primary section, except a button to re-enable it, when the disable primary is toggled on, and returned when it is toggled off?

The internal verdict was to remove entirely when primary is disabled, not even leaving a button to re-enable it. Beside entries for existing instances, only the "about" and "quit" would remain. When `client.primary-name` becomes non-empty again, the buttons would return.

**Comment by surahman** [comment]
@ricab I need some guidance on the GUI modifications. The decorator pattern is familiar because I have done GUI work in Java but I am not proficient with Qt.

> Additionally, the [hotkey](https://github.com/canonical/multipass/blob/main/src/client/gui/gui_cmd.cpp#L107) should not open a shell anymore. That may be easier to achieve by just checking if primary is disabled inside the action callback.

I believe I have handled this with a safety check in the `open_multipass_shell` routine for an empty instance name.

> The internal verdict was to remove entirely when primary is disabled, not even leaving a button to re-enable it. Beside entries for existing instances, only the "about" and "quit" would remain. When `client.primary-name` becomes non-empty again, the buttons would return.

I believe I have partially achieved this by tightening up the `handle_petenv_instance`. The only location I have found where the `petenv` related menu separator items are added is in the `create_actions` routine, and it is here that I have added a check. The tray icon is missing the `petenv` section on a clean launch now and having any instance will crash the GUI - something is awry.

File: src/client/gui/gui_cmd.cpp
Comment (ricab) [inline src/client/gui/gui_cmd.cpp:173]: I am not sure if this was leftover from earlier, but I don't think we need a button to disable primary.

File: src/client/gui/gui_cmd.cpp
Comment (ricab) [inline src/client/gui/gui_cmd.cpp:245]:
```suggestion
    const bool petenv_visibility = !MP_SETTINGS.get(petenv_key).isEmpty();
```

File: src/platform/client/client_platform_linux.cpp
Comment (ricab) [inline src/platform/client/client_platform_linux.cpp:27]: Could this check be moved to the call site, please? 

I believe it's the job of the GUI and not so much this util function to avoid opening a shell when primary is disabled (this function is better regarded as having a(n undocumented) pre-condition that the instance name is valid). 

This can have an assert instead, if you like.

**Review by ricab — COMMENTED** [review]
Hey @surahman, this is looking quite alright, I think you've got the right approach and the GUI seems to be doing all we need :smiley:

Only a couple of details inline, but +1 on both checking disabled primary for shell opening and updating button visibility.

**Comment by ricab** [comment]
@surahman, I failed to notice that open_shell was being given an empty instance name to open the primary shell. So my talk about preconditions and asserts mustn't have made much sense. I just pushed what I meant: passing the actual name to open shell (giving us a proper window title), documenting the precondition in the util, and asserting (part of it) in the implementation.

Among other details, I also switched to using the cached setting, to avoid reading from disk multiple times. It shouldn't go stale for more than a second and better consider the same thing throughout anyway. Let me know what you think.

bors try

**Comment by surahman** [comment]
This line caught me out initially, until I looked up how we can display a message for an `assert`, nice 😉 :
```c++
assert(!instance_name.isEmpty() && "Instance name cannot be empty");
```

It seems like what you have done is solid, but this line seems to still be passing in a `null` string on line 111:
```c++
mp::cli::platform::open_multipass_shell(QString());
```

**Comment by ricab** [comment]
> This line caught me out initially, until I looked up how we can display a message for an `assert`, nice wink :

Yeah, asserts are weird in C++, no slot for a message, so need that trick :sunglasses: 

> It seems like what you have done is solid, but this line seems to still be passing in a null string on line 111 in src/client/gui/gui_cmd.cpp:

It looks alright [here](https://github.com/canonical/multipass/pull/2102/files#diff-e058fbe77c4ffdb2c0015d1339a6b6f9ede02d61b8921b1d8b7f01f9652fa660R111), are you perhaps looking at an old commit?

**Comment by surahman** [comment]
> It looks alright [here](https://github.com/canonical/multipass/pull/2102/files#diff-e058fbe77c4ffdb2c0015d1339a6b6f9ede02d61b8921b1d8b7f01f9652fa660R111), are you perhaps looking at an old commit?

Yup, sorry, was looking at one of the earlier commits - looks good 👍🏽

**Comment by ricab** [comment]
Hmm bors?

bors try

**Comment by ricab** [comment]
Oh there's a conflict that needs fixing. @surahman would you be able to fix that? Thanks!

**Comment by ricab** [comment]
bors cancel

**Comment by ricab** [comment]
bors try

**Comment by Saviq** [comment]
bors cancel

**Comment by Saviq** [comment]
bors try-

**Comment by Saviq** [comment]
bors try

**Comment by ricab** [comment]
bors cancel

**Comment by ricab** [comment]
bors try-

**Comment by ricab** [comment]
bors try

**Comment by ricab** [comment]
@surahman I believe you have dropped some test code when fixing the conflicts. I didn't want to force-push into your repo (probably couldn't anyway) so I have pushed a [fixed merge](https://github.com/canonical/multipass/commit/4cb066e4ac1d9c18cd74f383ef600148d28c6bc8) in the branch [primary-null-proposed](https://github.com/surahman/multipass/compare/set_client.primary-name=null...canonical:primary-null-proposed).

You should be able to move this PR to it with the following steps:

1. `git checkout set_client.primary-name=null`
2. `git fetch <canonicals_multipass_remote>`
3. `git reset --hard <canonicals_multipass_remote>/primary-null-proposed`
4. `git push --force` (after confirming it's pointing at the new merge commit)

**Comment by surahman** [comment]
@ricab sorry about that, thanks for the fix 👍🏽.

**Comment by ricab** [comment]
Alright, let's try again.

bors try

**Review by ricab — APPROVED** [review]
Alright, let's merge this. Yeyyy :partying_face: Thanks again for your effort getting this done @surahman 

bors merge

**Comment by surahman** [comment]
@ricab you are most welcome! Thanks for all your patience, guidance, and assistance with this.

**Comment by ricab** [comment]
bors retry

**Comment by ricab** [comment]
bors retry

**Pull request merged**
